sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/ExecutorServiceParallelExecutor.java

```diff
@@ -0,0 +1,373 @@
+
+  private final ConcurrentMap<StepAndKey, TransformExecutorService> currentEvaluations;
+  private final ConcurrentMap<TransformExecutor<?>, Boolean> scheduledExecutors;
+
+  private final Queue<ExecutorUpdate> allUpdates;
+  private final BlockingQueue<VisibleExecutorUpdate> visibleUpdates;
+
+  private final CompletionCallback defaultCompletionCallback;
+  private Collection<AppliedPTransform<?, ?, ?>> rootNodes;
+
+  public static ExecutorServiceParallelExecutor create(
+      ExecutorService executorService,
+      Map<PValue, Collection<AppliedPTransform<?, ?, ?>>> valueToConsumers,
+      TransformEvaluatorRegistry registry,
+      InProcessEvaluationContext context) {
+    return new ExecutorServiceParallelExecutor(
+        executorService, valueToConsumers, registry, context);
+  }
+
+  private ExecutorServiceParallelExecutor(
+      ExecutorService executorService,
+      Map<PValue, Collection<AppliedPTransform<?, ?, ?>>> valueToConsumers,
+      TransformEvaluatorRegistry registry,
+      InProcessEvaluationContext context) {
+    this.executorService = executorService;
+    this.valueToConsumers = valueToConsumers;
+    this.registry = registry;
+    this.evaluationContext = context;
+
+    currentEvaluations = new ConcurrentHashMap<>();
+    scheduledExecutors = new ConcurrentHashMap<>();
+
+    this.allUpdates = new ConcurrentLinkedQueue<>();
+    this.visibleUpdates = new ArrayBlockingQueue<>(20);
+
+    defaultCompletionCallback = new TimerlessCompletionCallback();
+  }
+
+  @Override
+  public void start(Collection<AppliedPTransform<?, ?, ?>> roots) {
+    rootNodes = ImmutableList.copyOf(roots);
+    Runnable monitorRunnable = new MonitorRunnable();
+    executorService.submit(monitorRunnable);
+  }
+
+  @SuppressWarnings("unchecked")
+  public void scheduleConsumption(
+      AppliedPTransform<?, ?, ?> consumer,
+      CommittedBundle<?> bundle,
+      CompletionCallback onComplete) {
+    evaluateBundle(consumer, bundle, onComplete);
+  }
+
+  private <T> void evaluateBundle(
+      final AppliedPTransform<?, ?, ?> transform,
+      @Nullable final CommittedBundle<T> bundle,
+      final CompletionCallback onComplete) {
+    final StepAndKey stepAndKey = StepAndKey.of(transform, bundle == null ? null : bundle.getKey());
+    TransformExecutorService state =
+        getStepAndKeyExecutorService(stepAndKey, bundle == null ? true : !bundle.isKeyed());
+    TransformExecutor<T> callable =
+        TransformExecutor.create(registry, evaluationContext, bundle, transform, onComplete, state);
+    state.schedule(callable);
+  }
+
+  private void scheduleConsumers(CommittedBundle<?> bundle) {
+    for (AppliedPTransform<?, ?, ?> consumer : valueToConsumers.get(bundle.getPCollection())) {
+      scheduleConsumption(consumer, bundle, defaultCompletionCallback);
+    }
+  }
+
+  private TransformExecutorService getStepAndKeyExecutorService(
+      StepAndKey stepAndKey, boolean parallelizable) {
```

Comment:
Whoops, this is broken as-is

The behavior that ends up happening is that the choice between execution styles occurs based on the first bundle to arrive at the step from upstream, which is potentially based on some flattened PCollection (with only one input being downstream of a GroupByKey)

I've fixed that and also split this method.